### ISSUE TITLE
ccache-like caching

### ISSUE THREAD

**Issue by CyberShadow** [dragon-lang/rund#22]
ccache is a wrapper around C/C++ compilers and related tools which caches the result of compilation based on the file contents, so if you compile source code, then delete / redownload / unpack to the same location and recompile, the second compilation will be cached.

It would be useful to have something similar in rund, probably as a mode enabled with a switch. Its effect would be:

1. Include contents of files when calculating the cache directory
2. Stop looking at timestamps when deciding whether something should be rebuilt
3. Make the compilation target always be the cache directory, even with `-od/-of`, and then copy artifacts out if necessary.

ccache also has a maximum cache size limit (upon hitting which it starts deleting old results), but that can be managed outside of rund or added later.

What do you think?

**Comment by marler8997** [comment]
Using the contents would be more robust than using timestamps alone.  The only concern would be performance.

There's no way to get caching right 100% of the time.  A compilation could depend on many things like the current datetime, the kernel version, the temperature, the hardware installed, etc and there's no way rund could know every input that a program depends on.  So coming up with the right set of inputs to the hash is a balancing act between performance and correctness.

Any new feature that makes the caching work in more cases would have to outweigh the performance cost.  So I would try to identify the use cases this feature helps (i.e. the redownload use case you mentioned) and then implement it to see how it affects performance.

There may also be a way to use the timestamp sometimes and fallback to using the contents other times? I'd have to think the use cases though...

**Comment by CyberShadow** [comment]
I think that if you were to map all files that participate in the compilation to memory and use a fast hash, the performance overhead should be minimal. IO should not be impacted as the compiler will need to read the files anyway.

**Comment by marler8997** [comment]
The compiler would only need to read the files on a cache miss though, which happens less often than a cache hit.  So the "cache hit" case which is supposed to be the fast path would now need to read in all the files and hash them on every invocation.

**Comment by CyberShadow** [comment]
Right, but it would still be much much faster than actually invoking the compiler.

Further optimizing that, e.g. saving the files' hashes (reusing them if the timestamp hasn't changed) and then using something like a Merkle tree, could be a future optimization.

**Comment by marler8997** [comment]
> Right, but it would still be much much faster than actually invoking the compiler.

Sure but you wouldn't be invoking the compiler at all in the "cache hit" case.

Say you're project uses 100 files in compilation.  Checking the timestamps may take a few milliseconds.  If hashing all those files each time took half a second, that means you're adding half a second to every invocation with `rund`.  In this theoretical case that doesn't sound like a win.  And with the current implmeentation, the worst case is that a file's timestamp gets updated (which I think would happen very rarely) and you recompile everything.  But that's the rare case.  We don't want to make the rare cases faster at the cost of making the common cases slower.

**Comment by CyberShadow** [comment]
>  Checking the timestamps may take a few milliseconds. If hashing all those files each time took half a second, that means you're adding half a second to every invocation with `rund`.

There's no way that should take that much.

```console
[vladimir@home phobos]$ find -name '*.d' -exec cat '{}' '+' | time md5sum
dae1d87d186480d983772e6da371be64  -
0.03user 0.00system 0:00.05elapsed 58%CPU (0avgtext+0avgdata 3220maxresident)k
0inputs+0outputs (0major+166minor)pagefaults 0swaps
```

All of Phobos takes 0.03 seconds to hash with just one CPU core.

**Comment by marler8997** [comment]
30 milliseconds for how many files?

**Comment by CyberShadow** [comment]
That would be for the 427 `.d` files in my Phobos checkout right now.

**Comment by marler8997** [comment]
Ok let's look at a real project.  What was that project you had that was fairly big?

**Comment by CyberShadow** [comment]
I don't know what that would refer to, but it's 27 ms for vibe.d (288 `.d` files).

**Comment by marler8997** [comment]
You had a project that took like 4 to 5 seconds to compile.  I'd like to see how many files we would need to hash in that case.  All the files in the JSON that the compiler outputs.

**Comment by CyberShadow** [comment]
> You had a project that took like 4 to 5 seconds to compile.

Sorry, I don't remember what that is. If it's the project from the "D compilation is too slow and I'm forking the compiler" post, then that project is actually pretty small (about a dozen files), it just did a lot of metaprogramming.

**Comment by marler8997** [comment]
Might have been `digger`.

Cache Miss Case: 7.9s
Cache Hit Case: .009s

There are about 220 `.d` files in the json file. so might take about 15 milliseconds to hash them each time.

So the issue I'm seeing with this is, for real projects, it's more than doubling the time it takes to run the project in the "happy path".  15 milliseconds is completely negligible in the slow path but it can be a big deal in the fast path.  And the use case where this would save time seems to be pretty rare...?  Not to mention that most tools use timestamps so it's expected behavior, i.e. I would expect that running `touch myfile.d` would cause it to be rebuilt.

**Comment by CyberShadow** [comment]
> So the issue I'm seeing with this is, for real projects, it's more than doubling the time it takes to run the project in the "happy path".

If this is an issue, the implementation could use timestamps as the first check. If the timestamps haven't changed, we can safely assume that the contents and hashes haven't either. (It does mean saving the hashes somewhere so we know what the cache directory is, so extra complexity.)

>  And the use case where this would save time seems to be pretty rare...?

It should be about the same scenarios as for ccache, and I've been getting good mileage out of it.

>  Not to mention that most tools use timestamps so it's expected behavior, i.e. I would expect that running `touch myfile.d` would cause it to be rebuilt.

As with ccache, it should have the exact same effect as if it was rebuilt, except for things like `__DATE__`.

Anyway, the bigger problem for why this should probably be opt-in is that this will cause the cache directory to grow much much faster, i.e. every time the program is edited and built will create a new entry.

**Comment by marler8997** [comment]
> If this is an issue, the implementation could use timestamps as the first check. If the timestamps haven't changed, we can safely assume that the contents and hashes haven't either. (It does mean saving the hashes somewhere so we know what the cache directory is, so extra complexity.)

Ok that's true.  I would prefer a little more complexity over making the happy path slower.  If I were to implement this, this is how I would do it.

> It should be about the same scenarios as for ccache, and I've been getting good mileage out of it.

Can you elaborate a bit?  How often do you think a file's timestamp would be updated without it's contents in a real world project?  Since other tools like Make don't support this, I know developers imlpement functions that will pre-generate contents and only write them to a file if they detect they are different so as not to update the timestamp if the contents haven't changed (see the `updateIfChanged` function at https://github.com/dlang/dmd/blob/master/src/build.d#L1088 for one example)

> Anyway, the bigger problem for why this should probably be opt-in is that this will cause the cache directory to grow much much faster, i.e. every time the program is edited and built will create a new entry.

Oh well if this was opt-in then I wouldn't care about adding a bit of time to the happy path.  I would be totally fine with making this an opt-in feature.

**Comment by CyberShadow** [comment]
> Can you elaborate a bit?

Things like:

- Switching branches
- Bisecting
- Alternating debug and release builds (when putting the binaries at different paths is impractical)
- Adding temporary/test edits and then undoing them

generally involve going back to an earlier version of the source code / build settings.

> Since other tools like Make don't support this, I know developers imlpement functions that will pre-generate contents and only write them to a file if they detect they are different so as not to update the timestamp if the contents haven't changed (see the `updateIfChanged` function at https://github.com/dlang/dmd/blob/master/src/build.d#L1088 for one example)

This pattern must be used carefully - if the source is newer than the target but the build step does not update the target (because its contents is unchanged), then the build step will be run every time (there is no way to know when the target actually needs to be rebuilt because this information isn't saved).

**Comment by marler8997** [comment]
> Things like:
> ...

Thanks for listing those.  That helps.

So to be clear, I have no problems making this `opt-in`.  Do you have a command-line option in mind?  With an opt-in solution I'd also be OK with hashing every time.  If we wanted to make it on by default, then I would want to save timestamps and hashes in a file to make the happy-path faster and I'd have to think on whether or not we would really want it on by default.

**Comment by CyberShadow** [comment]
>  Do you have a command-line option in mind?

Maybe `--content-cache`, it doesn't really matter to me as I run `rund` through a wrapper.

>  If we wanted to make it on by default

I think that some kind of cache size limit should be a requirement for making it the default too. E.g. on Arch Linux, `/tmp` is mounted on RAM and it might become a growing memory hog over time, and on Windows the temporary directory isn't cleared except by the "disk cleanup tool" which must be invoked by the user.

**Comment by marler8997** [comment]
Maybe I missed something.  I was thinking of saving the timestamps and hashes which shouldn't take much more space.  Why would this feature necessitate a cache size limit?

**Comment by CyberShadow** [comment]
It would need to save the output (executable) of compilation for every input, i.e. hash of source files. D executables can be pretty big (over 1 MB) so it might grow quickly.

**Comment by marler8997** [comment]
Right now `rund` only saves the final executable (i.e. doesn't save any object files).  Changing this would be a big refactor of rund and I think lies outside the scope of rund.  To me that's sounds more like a feature for a build system.

I see `rund` as a tool to quickly run D executables, not really as a build system.  I would probably want another tool if I wanted to facilitate a system for working on large projects with quick rebuilds.

**Comment by CyberShadow** [comment]
>  Changing this would be a big refactor of rund and I think lies outside the scope of rund.

I don't see why this needs to change. The one thing that should change is that it should save the executable in its cache even when `-od`/`-of` are on the command line.

**Comment by marler8997** [comment]
Oh I misunderstood you.

We wouldn't need to save the executable in the cache in that case, we could just save it's hash.

**Comment by CyberShadow** [comment]
We will need to recreate it every time, so we would need to save a copy.

**Comment by marler8997** [comment]
> We will need to recreate it every time,...

I'm missing something, why is this?

**Comment by CyberShadow** [comment]
```console
$ echo 'int main() { return 0; }' > test.d
$ rund --content-cache --build-only -oftest test.d # full build
$ echo 'int main() { return 1; }' > test.d
$ rund --content-cache --build-only -oftest test.d # full build
$ echo 'int main() { return 0; }' > test.d
$ rund --content-cache --build-only -oftest test.d # should pull it from cache, where will "test" be copied from?
```

**Comment by marler8997** [comment]
Ok I see now.

We're trading disk space for supporting more cache cases.  I'm not sure I can say whether caching the executable in the `-of/-od` case is better or not in general.  This also seems like a rare use case to me, which could be solved by using different source files...
```
$ echo 'int main() { return 0; }' > true.d
$ rund --content-cache --build-only -oftrue true.d # full build
$ echo 'int main() { return 1; }' > false.d
$ rund --content-cache --build-only -offalse false.d # full build
$ echo 'int main() { return 0; }' > true.d
$ rund --content-cache --build-only -oftrue true.d # will not build because source content is the same and `true` executable timestamp/hash will be verified
```

**Comment by CyberShadow** [comment]
>  I'm not sure I can say whether caching the executable in the `-of/-od` case is better or not in general.

Yes, it's a tradeoff (copying a file after building vs. being able to quickly produce it from cache).

> which could be solved by generating different binaries

Yes, but as mentioned above it's not always practical. The executable of the D program might be part of a bigger pipeline and it might be impractical to change filenames there every time.

**Comment by marler8997** [comment]
> Yes, but as mentioned above it's not always practical. The executable of the D program might be part of a bigger pipeline and it might be impractical to change filenames there every time.

Can't really think of any situations where this applies.

**Comment by CyberShadow** [comment]
If the D program is part of a bigger system such as a makefile which then invokes it, then you would have to edit the makefile every time you want to use a different D executable file name (for cacheability).

**Comment by marler8997** [comment]
Could you provide an example?  I understand what you're saying but I'm not seeing a use-case where this would happen.

Keep in mind that this use case would also need a reason to use `-of/-od`.  In the use case you seem to be describing, why wouldn't you just use the `rund` cache?  And if it's because you want the executable to be under a certain path (i.e. the project directory), you could override the cache directory and each binary will still have a unique path based on the input hash.

```
# this allows rund to generate the unique binary path for you
rund --cache=. myfile.d
```

**Comment by CyberShadow** [comment]
> Could you provide an example?

One case which fits would be the `make && sudo make install` pattern. Here, it's not known if the executable is invoked in order to be invoked or in order to be copied somewhere, so the latter must be assumed.

**Comment by marler8997** [comment]
But in that case why would you be re-compiling the same executable filename with different sources?  I'm still not seeing a use case yet.

**Comment by CyberShadow** [comment]
> But in that case why would you be re-compiling the same executable filename with different sources?

If the program needs to be installed system-wide to be usable (e.g. the software using it is another user on the same machine, or the program can't be meaningfully run from its source directory).

The executable filename would be difficult to change as it involves messing with the Makefile.

The sources may change according to the situations I mentioned above (switching branches, etc.)

**Comment by marler8997** [comment]
This seems like a rare case.   You're choosing to save the executable in the cache every time the user requests `-of/-od` in order to support the case where a build system builds an executable, rebuilds the same executable with different inputs, then builds the same executable again with the original inputs.  I think saving only the most recent executable in the cache (which is what rund does today) support 90% of cases and only takes minimal disk space, it seems like a good balance to me.

And for the other 10% of cases that this wouldn't work for, there are ways to mitigate the issues using --cache or changing the executable name, etc.

I think it comes down to what's more likely.   If you're using `-of/-od`, is it more likely that you are going to need multiple past versions of the same executable, or do you just want to save the latest version of each executable?

**Comment by marler8997** [comment]
I should clarify.

I'm not saying that saving the executable in the cache with `-of/-od` is worse.  I'm saying I don't actually know which is better.  It's not a super high cost as `-of/-od` is less common.  I'm being more critical of this feature as it's a change from the status quo and I avoid change unless I can determine it's an improvement from the status quo.

**Comment by CyberShadow** [comment]
> This seems like a rare case. You're choosing to save the executable in the cache every time the user requests `-of/-od` in order to support the case where a build system builds an executable, rebuilds the same executable with different inputs, then builds the same executable again with the original inputs. I think saving only the most recent executable in the cache (which is what rund does today) support 90% of cases and only takes minimal disk space, it seems like a good balance to me.

Yes, and all of this makes sense, but this is still a feature that I would personally find useful.

> And for the other 10% of cases that this wouldn't work for, there are ways to mitigate the issues using --cache or changing the executable name, etc.

These workarounds don't work for me. E.g., I thought maybe I could make my rund wrapper swallow `-of`/`-od`, but then it doesn't have a way to find out with certainty the location where `rund` will place the executable in its cache to copy it out of there.

Maybe I should mention that I *always* use `-od.` (actually `--od="$(dirname "$prog")"`), it is completely ingrained in my workflow and scripts.

> I think it comes down to what's more likely. If you're using `-of/-od`, is it more likely that you are going to need multiple past versions of the same executable, or do you just want to save the latest version of each executable?

I suggest making rund save past versions of executables only with `--content-cache`.

**Comment by marler8997** [comment]
> These workarounds don't work for me. E.g., I thought maybe I could make my rund wrapper swallow -of/-od, but then it doesn't have a way to find out with certainty the location where rund will place the executable in its cache to copy it out of there.

Maybe we should make a way to access the cache binary location?

Instead of using `-of/-od`, we could add an option like `--cache-link <pathname>` which would cause rund to create a symbolic link to the cache location?

```
$ rund --cache-link foo foo.d
$ ./foo # runs the program
$ readlink foo
/tmp/.rund-1000/rund-foo.d-<HASH>/foo
```

You could also combine this with `--cache` to make sure the executable is still stored in your project directory.

**Comment by CyberShadow** [comment]
> Instead of using `-of/-od`, we could add an option like `--output-link <pathname>` or `-ol <pathname>` which would cause rund to create a symbolic link to the cache location?

I think that's a great idea. Too bad symlinks are a pain on Windows.

**Comment by marler8997** [comment]
> I think that's a great idea. Too bad symlinks are a pain on Windows.

https://github.com/dragon-lang/rund/issues/23

**Comment by CyberShadow** [comment]
Someone implemented this idea in a stand-alone tool:

https://github.com/seeraven/dmdcache
https://forum.dlang.org/post/r812of$11n7$1@digitalmars.com

Would be good to integrate it directly into a build tool like rund, so that it can be used for ad-hoc builds too.
